Add Progress Bar to docs (#11359)
PyTorchLightning/pytorch-lightning: docs/source/common/progress_bar.rst

.. testsetup:: *

    from pytorch_lightning.trainer.trainer import Trainer

> Progress Bar

Author: Updated!

============

Lightning supports two different types of progress bars (`tqdm <https://github.com/tqdm/tqdm>`_ and `rich <https://github.com/Textualize/rich>`_). :class:`~pytorch_lightning.callbacks.TQDMProgressBar` is used by default,
but you can override it by passing a custom :class:`~pytorch_lightning.callbacks.TQDMProgressBar` or :class:`~pytorch_lightning.callbacks.RichProgressBar` to the ``callbacks`` argument of the :class:`~pytorch_lightning.trainer.trainer.Trainer`.

Supported Progress Bars
-----------------------

.. currentmodule:: pytorch_lightning.callbacks.progress

.. autosummary::
    :toctree: generated
    :nosignatures:
    :template: classtemplate.rst

    RichProgressBar
    TQDMProgressBar

-------

> TQDMProgressBar

Reviewer: There's still technical limitations to do the switch right?

Keep it as it is and we can move it when it actually happens

.. testsetup:: *

    from pytorch_lightning.trainer.trainer import Trainer

Progress Bar
============

> Lightning supports two different types of progress bars (`tqdm <https://github.com/tqdm/tqdm>`_ and `rich <https://github.com/Textualize/rich>`_). :class:`~pytorch_lightning.callbacks.TQDMProgressBar` is used by default,

Reviewer 2: 1. should this mention the base progress bar class somewhere? 

or at this point, do we even need the base class? 
- progress tracking is done in the loop
- once https://github.com/PyTorchLightning/pytorch-lightning/issues/10982 is in, then enable/disable aren't required on the progress bar API

2. should we mention that a progress bar is automatically enabled with the Trainer, and to disable it, one should do this:`Trainer(enable_progress_bar=False)`

Author: Updated!

> but you can override it by passing custom :class:`~pytorch_lightning.callbacks.TQDMProgressBar` or :class:`~pytorch_lightning.callbacks.RichProgressBar` to the ``callbacks`` flag of :class:`~pytorch_lightning.trainer.trainer.Trainer`.

Reviewer 3: ```suggestion
but you can override it by passing a custom :class:`~pytorch_lightning.callbacks.TQDMProgressBar` or :class:`~pytorch_lightning.callbacks.RichProgressBar` to the ``callbacks`` flag of :class:`~pytorch_lightning.trainer.trainer.Trainer`.
```

Supported Progress Bars
-----------------------

.. currentmodule:: pytorch_lightning.callbacks.progress

.. autosummary::
    :toctree: generated
    :nosignatures:
    :template: classtemplate.rst

    RichProgressBar
    TQDMProgressBar

-------

TQDMProgressBar
---------------

> It uses the `tqdm <https://github.com/tqdm/tqdm>`_ library internally and is the default progress bar used by Lightning.

Reviewer 3: ```suggestion
The :class:`~pytorch_lightning.callbacks.TQDMProgressBar` uses the `tqdm <https://github.com/tqdm/tqdm>`_ library internally and is the default progress bar used by Lightning.
```

It prints to ``stdout`` and shows up to four different bars:

- **sanity check progress:** the progress during the sanity check run
- **main progress:** shows training + validation progress combined. It also accounts for multiple validation runs during training when :paramref:`~pytorch_lightning.trainer.trainer.Trainer.val_check_interval` is used.
- **validation progress:** only visible during validation; shows total progress over all validation datasets.
- **test progress:** only active when testing; shows total progress over all test datasets.

For infinite datasets, the progress bar never ends.

You can update ``refresh_rate`` (rate (number of batches) at which the progress bar get updated) for :class:`~pytorch_lightning.callbacks.TQDMProgressBar` by:

.. code-block:: python

    from pytorch_lightning.callbacks import TQDMProgressBar

>     trainer = Trainer(callbacks=TQDMProgressBar(refresh_rate=10))

Reviewer 2: should we use list convention everywhere when specifying callbacks on the trainer? 
```suggestion
    trainer = Trainer(callbacks=[TQDMProgressBar(refresh_rate=10)])
```

Reviewer: Why? Passing a single callback is supported.

Author: Yup.

Reviewer 2: Using the list reinforces that passing multiple callbacks is supported. This could avoid confusion for complete beginners who go through the docs and see single callbacks configured, but who aren't aware of how to chain them together. It also avoids confusion between `Trainer(callbacks=<Single class>)` having a different type than `trainer.callbacks`

If you want to customize the default :class:`~pytorch_lightning.callbacks.TQDMProgressBar` used by Lightning, you can override
specific methods of the callback class and pass your custom implementation to the :class:`~pytorch_lightning.trainer.trainer.Trainer`.

.. code-block:: python

    class LitProgressBar(TQDMProgressBar):
        def init_validation_tqdm(self):
            bar = super().init_validation_tqdm()
            bar.set_description("running validation...")
            return bar

    bar = LitProgressBar()

>     trainer = Trainer(callbacks=[bar])

Author: Updated!

.. seealso::
    - :class:`~pytorch_lightning.callbacks.TQDMProgressBar` docs.
    - `tqdm library <https://github.com/tqdm/tqdm>`__

----------------

RichProgressBar
---------------

`Rich <https://github.com/Textualize/rich>`_ is a Python library for rich text and beautiful formatting in the terminal.

> To use :class:`~pytorch_lightning.callbacks.RichProgressBar` as your progress bar do the following. First, install the package:

Reviewer 3: ```suggestion
To use the :class:`~pytorch_lightning.callbacks.RichProgressBar` as your progress bar, first install the package:
```

.. code-block:: bash

    pip install rich

Then configure the callback and pass it to the :class:`~pytorch_lightning.trainer.trainer.Trainer`:

.. code-block:: python

    from pytorch_lightning.callbacks import RichProgressBar

>     trainer = Trainer(callbacks=RichProgressBar())

Reviewer 2: ```suggestion
    trainer = Trainer(callbacks=[RichProgressBar()])
```

> Customize the theme for your :class:`~pytorch_lightning.callbacks.RichProgressBar` by:

Reviewer 3: ```suggestion
Customize the theme for your :class:`~pytorch_lightning.callbacks.RichProgressBar` like this:
```

.. code-block:: python

    from pytorch_lightning.callbacks import RichProgressBar

>     from pytorch_lightning.callbacks.progress.rich_progress import RichProgressBarTheme

Reviewer 2: i'm not sure if our naming is consistent here w/ other directories. should we rename this file to be
`pytorch_lightning.callbacks.progress.rich` ?

Reviewer 4: Its directory could also be renamed from `pytorch_lightning.callbacks.progress` to `pytorch_lightning.callbacks.progress_bars` which I think sounds more natural.

    # create your own theme!
    progress_bar = RichProgressBar(
        theme=RichProgressBarTheme(
            description="green_yellow",
            progress_bar="green1",
            progress_bar_finished="green1",
            progress_bar_pulse="#6206E0",
            batch_progress="green_yellow",
            time="grey82",
            processing_speed="grey82",
            metrics="grey82",
        )
    )

>     trainer = Trainer(callbacks=progress_bar)

Reviewer 2: ```suggestion
    trainer = Trainer(callbacks=[progress_bar])
```

> You could customize the components used within :class:`~pytorch_lightning.callbacks.RichProgressBar` with ease by overriding the

Reviewer 3: ```suggestion
You can customize the components used within :class:`~pytorch_lightning.callbacks.RichProgressBar` with ease by overriding the
```

:func:`~pytorch_lightning.callbacks.RichProgressBar.configure_columns` method.

.. code-block:: python

    from rich.progress import TextColumn

    custom_column = TextColumn("[progress.description]Custom Rich Progress Bar!")

    class CustomRichProgressBar(RichProgressBar):
        def configure_columns(self, trainer):
            return [custom_column]

    progress_bar = CustomRichProgressBar()

> If you wish a new progress bar to be displayed at the end of every epoch, you could enable

Reviewer 3: ```suggestion
If you wish for a new progress bar to be displayed at the end of every epoch, you should enable
```

> :paramref:`~pytorch_lightning.callbacks.RichProgressBar.leave` by passing ``True``

Reviewer 3: ```suggestion
:paramref:`RichProgressBar.leave <pytorch_lightning.callbacks.RichProgressBar.leave>` by passing ``True``
```
